controller/src/test/java/io/pravega/controller/server/v1/ControllerServiceImplTest.java

```diff
@@ -304,6 +306,46 @@ public void deleteStreamTests() {
         assertEquals("Delete sealed stream", DeleteStreamStatus.Status.SUCCESS, deleteStreamStatus.getStatus());
     }
 
+    @Test
+    public void streamCutValidationTest() {
+        // call scale test to create stream and scale it
+        scaleTest();
+
+        // Case 1: do not cover full range
+        ResultObserver<Controller.StreamCutValidityResponse> result1 = new ResultObserver<>();
+        StreamInfo streamInfo = StreamInfo.newBuilder()
+                .setScope(SCOPE1)
+                .setStream(STREAM1)
+                .build();
+        this.controllerService.isStreamCutValid(Controller.StreamCut.newBuilder()
+                .setStreamInfo(streamInfo)
+                .putCut(0, 0).build(), result1);
+        assertFalse(result1.get().getResponse());
+
+        // Case 2: include overlapping segments
+        ResultObserver<Controller.StreamCutValidityResponse> result2 = new ResultObserver<>();
+
+        this.controllerService.isStreamCutValid(Controller.StreamCut.newBuilder()
+                .setStreamInfo(streamInfo).putCut(0, 0).putCut(1, 0).
+                        putCut(computeSegmentId(2, 1), 0).
+                        putCut(computeSegmentId(3, 1), 0).build(), result2);
+        assertFalse(result2.get().getResponse());
+
+        // Case 3: Correct stream cut spanning one epoch
+        ResultObserver<Controller.StreamCutValidityResponse> result3 = new ResultObserver<>();
+        this.controllerService.isStreamCutValid(Controller.StreamCut.newBuilder()
+                .setStreamInfo(streamInfo).putCut(0, 0).putCut(1, 0).build(), result3);
+        assertTrue(result3.get().getResponse());
+
+        // Case 4: Correct stream cut spanning two epochs
+        ResultObserver<Controller.StreamCutValidityResponse> result4 = new ResultObserver<>();
+        this.controllerService.isStreamCutValid(Controller.StreamCut.newBuilder()
+                .setStreamInfo(streamInfo).putCut(0, 0).
+                putCut(computeSegmentId(2, 1), 0).
+                putCut(computeSegmentId(3, 1), 0).build(), result4);
+        assertTrue(result4.get().getResponse());
```

Comment:
We should also cover incorrect cuts spanning two epocs. For example given the following segment history:
1,2 => 3 => 4,5
then a cut of 1,5 is invalid even if they perfectly evenly split keyspace.

Reply:
This PR is not changing the validation logic, merely adding a new api. We will address validation as part of separate PR after having finalized the representation.